File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/infrastructure/zlink/actors/SupportActorDirectory.java
package systems.zlink.samples.supportchat.server.support.infrastructure.zlink.actors;

import java.util.HashMap;
import java.util.Map;
import systems.zlink.framework.actors.ZLinkActorRef;

public final class SupportActorDirectory {
    public record Entry(ZLinkActorRef ref, String displayName, String role) {
    }

    private final Object gate = new Object();
    private final Map<String, Entry> actors = new HashMap<>();

    public void addOrUpdate(ZLinkActorRef ref, String displayName, String role) {
        synchronized (gate) {
            actors.put(ref.actorId(), new Entry(ref, displayName, role));
        }
    }

    public Entry get(String actorId) {
        synchronized (gate) {
            Entry entry = actors.get(actorId);
            if (entry == null) {
                throw new IllegalStateException("Support actor is not available. actor=" + actorId);
            }
            return entry;
        }
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/actors/SupportActorDirectory.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors

import systems.zlink.framework.actors.ZLinkActorRef

class SupportActorDirectory {
    data class Entry(val ref: ZLinkActorRef, val displayName: String, val role: String)

    private val gate = Any()
    private val actors = mutableMapOf<String, Entry>()

    fun addOrUpdate(ref: ZLinkActorRef, displayName: String, role: String) {
        synchronized(gate) {
            actors[ref.actorId()] = Entry(ref, displayName, role)
        }
    }

    fun get(actorId: String): Entry =
        synchronized(gate) {
            actors[actorId]
                ?: throw IllegalStateException("Support actor is not available. actor=$actorId")
        }
}
